src/main/java/org/broadinstitute/hellbender/tools/spark/RevertSamSpark.java

```diff
@@ -0,0 +1,706 @@
+
+        ////////////////////////////////////////////////////////////////////////////
+        // Construct appropriate headers for the output files
+        ////////////////////////////////////////////////////////////////////////////
+        final Map<String, SAMFileHeader> headerMap = getReadgroupHeaderMap(inHeader, writerMap);
+
+        // Revert the reads based on the given attributes
+        List<String> attributesToRevert = removeDefaults? DEFAULT_ATTRIBUTES_TO_CLEAR: new ArrayList<>();
+        attributesToRevert.addAll(attributesToClear);
+        JavaRDD<GATKRead> readsReverted = revertReads(reads, attributesToRevert);
+
+        ////////////////////////////////////////////////////////////////////////////
+        // Sanitize the reads, sorting them into appropriate order if necessary
+        ////////////////////////////////////////////////////////////////////////////
+        if (sanitize) {
+            Map<String, FastqQualityFormat> readGroupFormatMap = createReadGroupFormatMap(readsReverted, headerb, !dontRestoreOriginalQualities);
+
+            readsReverted = sanitize(readGroupFormatMap, readsReverted, inHeader, keepFirstDuplicate);
+        }
+
+        // Write the one or many read output files
+        for (Map.Entry<String, File> rmap: writerMap.entrySet()) {
+            //TODO what to do if the readgroup isn't present
+            final String key = rmap.getKey();
+            JavaRDD<GATKRead> filteredreads = rmap.getKey()==null? readsReverted :
+                                                                    readsReverted.filter(r -> r.getReadGroup().equals(key));
+            writeReads(ctx, rmap.getValue().getPath(), filteredreads, headerMap.get(rmap.getKey())); //TODO proper header map
+        }
+    }
+
+    /**
+     * Runs the QualityEncodingDetector over a sampling of each readGroup present in the file to detect what the encoding format
+     * for base quality those reads are.
+     *
+     * @param reads Reads RDD over which to iterate and detect readgroups
+     * @param inHeader Header describing the readgroups present in the bam
+     * @param restoreOriginalQualities Whether to use the OQ tag for determining the map
+     * @return
+     */
+    private Map<String, FastqQualityFormat> createReadGroupFormatMap( final JavaRDD<GATKRead> reads,
+                                                                      final Broadcast<SAMFileHeader> inHeader,
+                                                                      final boolean restoreOriginalQualities) {
+        final Map<String, FastqQualityFormat> output = new HashMap<>();
+
+        inHeader.getValue().getReadGroups().stream().parallel().forEach(rg -> {
+            // For each readgroup filter down to just the reads in that group
+            final String key = rg.getId();
+            JavaRDD<GATKRead> filtered = reads.filter(r -> r.getReadGroup().equals(key));
+
+            if (!filtered.isEmpty()) {
```

Comment:
This method seems like it could be very expensive.  Particularly in the case where you have very unbalanced distributions of reads.  (Worst case, many readgroups).  In that case I think you will end up iterating the whole file N-1 times.  I don't know how isEmpty works, but it seems like in cases where the rdd IS empty it would have to traverse the entire input and run the filter to prove that it is empty.

Have you noticed this taking a long time?

Reply:
I'm not sure what a clean alternative is... I've often wished for a partition operation on an RDD.  You can sort of do it with a groupByKey but it will have nasty out of memory issues...